File: 543-diameter-of-binary-tree/diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let max = 0

    const dfs = (root) => {
        if (root == null) return 0

        let left = dfs(root.left)
        let right = dfs(root.right)

        max = Math.max( max,left + right)

        return 1 + Math.max(left, right)
    }

    dfs(root)

    return max
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.m = 0
        def dfs(root):
            if root is None:
                return 0
            
            left = dfs(root.left)
            right = dfs(root.right)

            self.m = max(self.m, left + right)

            return 1 + max(left, right)
        
        dfs(root)
        return self.m